dev_docs/cxg.md

```diff
@@ -73,38 +70,38 @@ For example:
 }
 ```
 
-### X
+### `X` array
 
-TileDB array, with a single anonymous attribute, shape (n_obs, n_var), containing the count matrix (equivalent to the AnnData `X` array).  MUST have numeric type, and SHOULD be float32.  The TileDB schema defines type and sparsity, and both dense and sparse encoding are supported.
+TileDB array, with a single anonymous attribute, shape `(n_obs, n_var)`, containing the count matrix (equivalent to the AnnData `X` array).  MUST have numeric type, and SHOULD be `float32`.  The TileDB schema defines type and sparsity, and both dense and sparse encoding are supported.
 
-### X_col_shift
+### `X_col_shift` array
 
-Optional TileDB array, used to encode-per column offsets for column-shift sparse encoding.  The TileDB array will have a single anonymous attribute, of the same type as the X array, and shape (n_var,).  
+Optional TileDB array, used to encode per-column offsets for column-shift sparse encoding.  The TileDB array will have a single anonymous attribute, of the same type as the X array, and shape `(n_var,)`.  
 
-If the X array is sparse, and X_col_shift exists, then all values in the i'th column were subtracted by X_col_shift[i].
+If the `X` array is sparse, and the `X_col_shift` array exists, then all values in the `i`'th column were subtracted by `X_col_shift[i]`.
```

Comment:
why is this unique to sparse X matrix?  Seems like it is applicable to sparse or dense X matrix.

Reply:
I interpret the [h5ad -> cxg implementation](https://github.com/chanzuckerberg/single-cell-data-portal/blob/d9c534968d592aa9ecf46791b111e1e6edc061ba/backend/corpora/dataset_processing/h5ad_data_file.py#L99-L111) as follows: "If the H5AD-provided `X` matrix is not sufficiently sparse, per the desired `sparse_threshold`, attempt to compute a `col_shift` array that _would_ allow it to be stored as a sparse matrix that satisfies the desired `sparse_threshold`." Does this agree with your interpretation?

If so, perhaps the spec should read:
> If the `X_col_shift` array exists, then the matrix SHOULD be a TileDB sparse matrix, and all the values in the `i`'th column MUST be add by `X_col_shift[i]` to restore the originally-provided values from the source H5AD file.

I assume one could specify a `X_col_shift` in the CXG file even when the X matrix is stored densely, though that would be silly. Thus the `SHOULD`, above.

Reply:
I am not sure this spec should mandate a sparsity encoding threshold, or even get into the (policy) decision of when the matrix is sparse.  So, from the viewpoint of this spec, I think it would be more accurate to say a small variation on your original theme:

> If the `X` array is sparse, and the `X_col_shift` array exists, then all values in the `i`'th column of `X` are decremented by the value of the corresponding `X_col_shift[i]`.   `X_col_shift` MAY exist if the `X` array is sparse, and when it doe not exist, has identical semantics to a column shift value of zero (0) for each column.   `X_col_shift` MUST NOT exist if the `X` array is dense.

TL;DR:
* dense must not have a col shift array
* sparse may have a col-shift array, and if it doesn't, it is identical to the semantics of a col-shift of zero

Reply:
agreed, the sparsity threshold is not a spec concern, which should only care about ensuring the `X_col_shift` is used correctly.

will incorporate your rewrite, which additionally stipulates "dense implies no x_col_shift"